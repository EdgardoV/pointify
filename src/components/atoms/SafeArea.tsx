import React, {ReactNode} from 'react';
import {
  View,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  ViewStyle,
  StatusBarStyle,
  Animated,
} from 'react-native';
import Theme from 'theme';
import {useColorScheme} from 'hooks/useColorScheme';

type BarStyle = 'light' | 'dark';

interface Props {
  topSafeArea?: boolean;
  bottomSafeArea?: boolean;
  safeBGColor?: ViewStyle['backgroundColor'];
  bottomBGColor?: ViewStyle['backgroundColor'];
  topBGColor?: ViewStyle['backgroundColor'];
  children: ReactNode;
  barStyle?: BarStyle;
  backgroundColor?: ViewStyle['backgroundColor'];
  statusBarColor?: ViewStyle['backgroundColor'];
  hiddenStatusBar?: boolean;
}

const SafeArea: React.FC<Props> = ({
  topSafeArea = true,
  bottomSafeArea = true,
  safeBGColor,
  children,
  barStyle,
  backgroundColor,
  bottomBGColor,
  topBGColor,
  statusBarColor,
  hiddenStatusBar,
}: Props) => {
  const {colorScheme} = useColorScheme();

  const renderChildren = () => (
    <Animated.View
      style={[
        styles.childrenContainerStyle,
        {
          backgroundColor:
            backgroundColor || Theme.Colors.Background[colorScheme],
        },
      ]}>
      {children}
    </Animated.View>
  );

  const renderStatusBar = () => {
    let barStyleValue: StatusBarStyle = 'default';

    switch (barStyle) {
      case 'light':
        barStyleValue = 'light-content';
        break;
      case 'dark':
        barStyleValue = 'dark-content';
        break;
      default:
        barStyleValue =
          colorScheme === 'dark' ? 'light-content' : 'dark-content';
        break;
    }

    return (
      <StatusBar
        translucent={!topSafeArea}
        backgroundColor={
          !topSafeArea
            ? statusBarColor || 'transparent'
            : topBGColor || safeBGColor || Theme.Colors.Background[colorScheme]
        }
        barStyle={barStyleValue}
        hidden={hiddenStatusBar}
      />
    );
  };

  const renderComponent = () => (
    <View style={styles.flexStyle}>
      {renderStatusBar()}
      {renderChildren()}
    </View>
  );

  const renderContainer = () => (
    <View style={styles.flexStyle}>
      {topSafeArea && (
        <SafeAreaView
          style={{
            backgroundColor:
              topBGColor || safeBGColor || Theme.Colors.Background[colorScheme],
          }}
        />
      )}
      {renderComponent()}
      {bottomSafeArea && (
        <SafeAreaView
          style={{
            backgroundColor:
              bottomBGColor ||
              safeBGColor ||
              Theme.Colors.Background[colorScheme],
          }}
        />
      )}
    </View>
  );

  return <View style={styles.flexStyle}>{renderContainer()}</View>;
};

const styles = StyleSheet.create({
  flexStyle: {
    flex: 1,
  },
  childrenContainerStyle: {
    flex: 1,
  },
});

export {SafeArea};
