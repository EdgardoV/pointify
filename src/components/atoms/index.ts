export * from './CustomText';
export * from './SafeArea';
export * from './TextSpan';
