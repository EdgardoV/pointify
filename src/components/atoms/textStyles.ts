import {StyleSheet} from 'react-native';
import Theme from '../../theme';

const typographyLightStyle = StyleSheet.create({
  header: {
    fontSize: Theme.Sizes.Heading,
    fontFamily: Theme.Fonts.Medium,
    color: Theme.Colors.PrimaryText.light,
  },
  title: {
    fontSize: Theme.Sizes.Title,
    fontFamily: Theme.Fonts.Medium,
    color: Theme.Colors.PrimaryText.light,
  },
  subtitle: {
    fontFamily: Theme.Fonts.Bold,
    fontSize: Theme.Sizes.Subtitle,
    color: Theme.Colors.PrimaryText.light,
  },
  paragraph: {
    fontSize: Theme.Sizes.Paragraph,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.light,
  },
  caption1: {
    fontSize: Theme.Sizes.Caption1,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.light,
  },
  caption2: {
    fontSize: Theme.Sizes.Caption2,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.light,
  },
});

const typographyDarkStyle = StyleSheet.create({
  header: {
    fontSize: Theme.Sizes.Heading,
    fontFamily: Theme.Fonts.Medium,
    color: Theme.Colors.PrimaryText.dark,
  },
  title: {
    fontSize: Theme.Sizes.Title,
    fontFamily: Theme.Fonts.Medium,
    color: Theme.Colors.PrimaryText.dark,
  },
  subtitle: {
    fontFamily: Theme.Fonts.Regular,
    fontSize: Theme.Sizes.Subtitle,
    color: Theme.Colors.PrimaryText.dark,
  },
  paragraph: {
    fontSize: Theme.Sizes.Paragraph,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.dark,
  },
  caption1: {
    fontSize: Theme.Sizes.Caption1,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.dark,
  },
  caption2: {
    fontSize: Theme.Sizes.Caption2,
    fontFamily: Theme.Fonts.Regular,
    color: Theme.Colors.PrimaryText.dark,
  },
});

export {typographyLightStyle, typographyDarkStyle};
