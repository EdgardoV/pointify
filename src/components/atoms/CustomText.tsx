import React from 'react';
import {Text, TextStyle, StyleProp} from 'react-native';
import {typographyDarkStyle, typographyLightStyle} from './textStyles';
import {useColorScheme} from 'hooks/useColorScheme';
import Theme from 'theme';

export type TypographyTypes =
  | 'header'
  | 'title'
  | 'subtitle'
  | 'paragraph'
  | 'caption1'
  | 'caption2'
  | 'caption3'
  | 'caption4';

export type FontWeightTypes = 'Regular' | 'Medium' | 'Bold' | 'Light';

export interface CustomTextProps {
  text: string;
  textAlign?: TextStyle['textAlign'];
  textColor?: TextStyle['color'];
  typography?: TypographyTypes;

  fontWeight?: FontWeightTypes;
  fontSize?: number;
  numberOfLines?: number;
  transform?: TextStyle['textTransform'];
  underline?: boolean;
  allowFontScaling?: boolean;
  spacing?: number;
  lineHeight?: number;
}

const CustomText: React.FC<CustomTextProps> = ({
  text,
  typography = 'paragraph',
  textColor,
  textAlign,
  fontWeight,
  fontSize,
  numberOfLines,
  transform,
  underline,
  allowFontScaling,
  spacing,
  lineHeight,
}: CustomTextProps) => {
  const {colorScheme} = useColorScheme();

  const style =
    colorScheme === 'dark' ? typographyDarkStyle : typographyLightStyle;

  const textStyle: StyleProp<TextStyle> = [];
  switch (typography) {
    case 'header':
      textStyle.push(style.header);
      break;
    case 'title':
      textStyle.push(style.title);
      break;
    case 'subtitle':
      textStyle.push(style.subtitle);
      break;
    case 'paragraph':
      textStyle.push(style.paragraph);
      break;
    case 'caption1':
      textStyle.push(style.caption1);
      break;
    case 'caption2':
      textStyle.push(style.caption2);
      break;
    default:
      break;
  }

  if (fontSize) {
    textStyle.push({fontSize});
  }
  if (textColor) {
    textStyle.push({color: textColor});
  }

  if (fontWeight) {
    let selectedWeight: FontWeightTypes = 'Regular';

    if (fontWeight) {
      selectedWeight = fontWeight;
    } else {
      switch (typography) {
        case 'caption1':
          selectedWeight = 'Regular';
          break;
        case 'caption2':
          selectedWeight = 'Regular';
          break;
        case 'caption4':
          selectedWeight = 'Regular';
          break;
        case 'header':
          selectedWeight = 'Medium';
          break;
        case 'paragraph':
          selectedWeight = 'Regular';
          break;
        case 'subtitle':
          selectedWeight = 'Regular';
          break;
        case 'title':
          selectedWeight = 'Medium';
          break;
        default:
          break;
      }
    }

    textStyle.push({
      fontFamily: Theme.Fonts[selectedWeight],
    });
  }

  return (
    <Text
      style={[
        textStyle,
        // eslint-disable-next-line react-native/no-inline-styles
        {
          textAlign,
          textTransform: transform,
          textDecorationLine: underline ? 'underline' : 'none',
          letterSpacing: spacing,
          lineHeight,
        },
      ]}
      ellipsizeMode="tail"
      numberOfLines={numberOfLines}
      allowFontScaling={allowFontScaling}>
      {text}
    </Text>
  );
};

export {CustomText};
