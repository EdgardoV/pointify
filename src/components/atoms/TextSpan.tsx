import React, {ReactNode} from 'react';
import {StyleProp, TextStyle, Text} from 'react-native';
import {useColorScheme} from 'hooks';
import Theme from 'theme';

import {typographyDarkStyle, typographyLightStyle} from './textStyles';
import {FontWeightTypes, TypographyTypes} from './CustomText';

export interface TextSpanProps {
  text?: string;
  textAlign?: TextStyle['textAlign'];
  textColor?: TextStyle['color'];
  typography?: TypographyTypes;

  fontWeight?: FontWeightTypes;

  marginRight?: number;
  marginLeft?: number;
  marginTop?: number;
  marginBottom?: number;
  fontSize?: number;

  children?: ReactNode;
  onPress?: () => void;

  transform?: TextStyle['textTransform'];
  underline?: boolean;
  spacing?: number;
  lineHeight?: number;
}

const TextSpan: React.FC<TextSpanProps> = ({
  text = '',
  typography = 'paragraph',
  textColor,
  textAlign,
  fontWeight,
  children,
  onPress,
  fontSize,
  marginTop,
  marginBottom,
  marginLeft,
  marginRight,
  transform,
  underline,
  spacing,
  lineHeight,
}) => {
  const {colorScheme} = useColorScheme();

  const style =
    colorScheme === 'dark' ? typographyDarkStyle : typographyLightStyle;
  const textStyle: StyleProp<TextStyle> = [];

  const defineTypography = () => {
    switch (typography) {
      case 'header':
        textStyle.push(style.header);
        break;
      case 'title':
        textStyle.push(style.title);
        break;
      case 'subtitle':
        textStyle.push(style.subtitle);
        break;
      case 'paragraph':
        textStyle.push(style.paragraph);
        break;
      case 'caption1':
        textStyle.push(style.caption1);
        break;
      case 'caption2':
        textStyle.push(style.caption2);
        break;
      default:
        break;
    }
  };

  if (children) {
    defineTypography();

    textStyle.push({
      color: textColor || Theme.Colors.PrimaryText[colorScheme],
      textAlign: textAlign || 'auto',
    });
  } else {
    if (typography) {
      defineTypography();
    }
    if (textColor) {
      textStyle.push({color: textColor});
    }
    if (textAlign) {
      textStyle.push({textAlign});
    }
  }

  if (fontSize) {
    textStyle.push({fontSize});
  }

  if (fontWeight) {
    let selectedWeight: FontWeightTypes = 'Regular';

    if (fontWeight) {
      selectedWeight = fontWeight;
    } else {
      switch (typography) {
        case 'caption1':
          selectedWeight = 'Regular';
          break;
        case 'caption2':
          selectedWeight = 'Regular';
          break;
        case 'header':
          selectedWeight = 'Medium';
          break;
        case 'paragraph':
          selectedWeight = 'Regular';
          break;
        case 'subtitle':
          selectedWeight = 'Regular';
          break;
        case 'title':
          selectedWeight = 'Medium';
          break;
        default:
          break;
      }
    }

    textStyle.push({
      fontFamily: Theme.Fonts[selectedWeight],
    });
  }

  textStyle.push({
    marginBottom,
    marginTop,
    marginLeft,
    marginRight,
    textTransform: transform,
    textDecorationLine: underline ? 'underline' : 'none',
    letterSpacing: spacing,
    lineHeight,
  });

  return (
    <Text onPress={onPress} style={textStyle}>
      {children || text}
    </Text>
  );
};

export {TextSpan};
