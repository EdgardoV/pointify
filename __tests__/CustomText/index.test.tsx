import React from 'react';
import renderer from 'react-test-renderer';
import {CustomText, CustomTextProps} from 'components/atoms/CustomText';

describe('CustomText', () => {
  const props: CustomTextProps = {
    text: 'Test CustomText',
    typography: 'paragraph',
    textColor: '#000000',
    textAlign: 'left',
    fontWeight: 'Regular',
    fontSize: 16,
    numberOfLines: 1,
    transform: 'none',
    underline: false,
    allowFontScaling: true,
    spacing: 0,
    lineHeight: 20,
  };

  it('renders correctly with props', () => {
    const tree = renderer.create(<CustomText {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without props', () => {
    const tree = renderer
      .create(<CustomText text="Test CustomText" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
