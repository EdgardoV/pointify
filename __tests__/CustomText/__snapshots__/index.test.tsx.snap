// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomText renders correctly with props 1`] = `
<Text
  allowFontScaling={true}
  ellipsizeMode="tail"
  numberOfLines={1}
  style={
    [
      [
        {
          "color": "#020202",
          "fontFamily": "Avenir",
          "fontSize": 12,
        },
        {
          "fontSize": 16,
        },
        {
          "color": "#000000",
        },
        {
          "fontFamily": "Avenir",
        },
      ],
      {
        "letterSpacing": 0,
        "lineHeight": 20,
        "textAlign": "left",
        "textDecorationLine": "none",
        "textTransform": "none",
      },
    ]
  }
>
  Test CustomText
</Text>
`;

exports[`CustomText renders correctly without props 1`] = `
<Text
  ellipsizeMode="tail"
  style={
    [
      [
        {
          "color": "#020202",
          "fontFamily": "Avenir",
          "fontSize": 12,
        },
      ],
      {
        "letterSpacing": undefined,
        "lineHeight": undefined,
        "textAlign": undefined,
        "textDecorationLine": "none",
        "textTransform": undefined,
      },
    ]
  }
>
  Test CustomText
</Text>
`;
